File: js/towerManager.js
import { TOWER_DEFINITIONS } from "./data/towerConfig.js";

export class TowerManager {
  constructor(game) {
    this.game = game;
    this.towers = [];
    this.projectiles = [];

    // We no longer embed tower data here; we import from TOWER_DEFINITIONS
    this.towerTypes = TOWER_DEFINITIONS;
  }

  getTowerData() {
    return this.towerTypes;
  }

  createTower(towerTypeName) {
    const def = this.towerTypes.find(t => t.type === towerTypeName);
    if (!def) return null;

    const firstLvl = def.upgrades[0];
    return {
      type: def.type,
      level: 1,
      range: def.range,
      damage: firstLvl.damage,
      splashRadius: def.splashRadius,
      fireRate: def.fireRate,
      fireCooldown: 0,
      upgradeCost: def.upgrades[1] ? def.upgrades[1].upgradeCost : 0,
      maxLevel: def.upgrades.length,
      x: 0,
      y: 0,
      spot: null,
      goldSpent: def.basePrice,
    };
  }

  update(deltaSec) {
    if (this.game.gameOver) return;

    // Fire towers
    this.towers.forEach(tower => {
      tower.fireCooldown -= deltaSec;
      if (tower.fireCooldown <= 0) {
        this.fireTower(tower);
        tower.fireCooldown = tower.fireRate;
      }
    });

    // Move projectiles
    this.projectiles.forEach(proj => {
      const step = proj.speed * deltaSec;
      const dx = proj.targetX - proj.x;
      const dy = proj.targetY - proj.y;
      const dist = Math.sqrt(dx * dx + dy * dy);

      if (dist <= step) {
        proj.x = proj.targetX;
        proj.y = proj.targetY;
        proj.hit = true;
      } else {
        proj.x += (dx / dist) * step;
        proj.y += (dy / dist) * step;
      }
    });

    // Handle collisions
    this.projectiles.forEach(proj => {
      if (proj.hit) {
        if (proj.splashRadius > 0) {
          // Splash damage
          const enemiesHit = this.game.enemies.filter(e => {
            const ex = e.x + e.width / 2;
            const ey = e.y + e.height / 2;
            const dx = proj.targetX - ex;
            const dy = proj.targetY - ey;
            return dx*dx + dy*dy <= proj.splashRadius * proj.splashRadius;
          });
          enemiesHit.forEach(e => {
            if (e === proj.mainTarget) e.hp -= proj.damage;
            else e.hp -= proj.damage / 2;
          });
        } else {
          // Single target
          if (proj.mainTarget) {
            proj.mainTarget.hp -= proj.damage;
          }
        }
      }
    });

    // Clean up projectiles that have hit
    this.projectiles = this.projectiles.filter(p => !p.hit);
  }

  fireTower(tower) {
    const enemiesInRange = this.game.enemies.filter(e => {
      const ex = e.x + e.width / 2;
      const ey = e.y + e.height / 2;
      const dx = ex - tower.x;
      const dy = ey - tower.y;
      return (dx*dx + dy*dy) <= (tower.range * tower.range);
    });
    if (!enemiesInRange.length) return;

    // Lock onto first enemy
    const target = enemiesInRange[0];
    const ex = target.x + target.width / 2;
    const ey = target.y + target.height / 2;

    this.projectiles.push({
      x: tower.x,
      y: tower.y,
      w: 4,
      h: 4,
      speed: 300,
      damage: tower.damage,
      splashRadius: tower.splashRadius,
      mainTarget: target,
      targetX: ex,
      targetY: ey,
      hit: false,
    });
  }

  upgradeTower(tower) {
    const def = this.towerTypes.find(t => t.type === tower.type);
    if (!def) return;
    if (tower.level >= def.upgrades.length) return;

    const nextLvlIndex = tower.level;
    const nextLvl = def.upgrades[nextLvlIndex];
    if (!nextLvl) return;
    if (this.game.gold < nextLvl.upgradeCost) return;

    // Spend gold
    this.game.gold -= nextLvl.upgradeCost;
    tower.goldSpent += nextLvl.upgradeCost;
    tower.level++;

    tower.damage = nextLvl.damage;
    tower.upgradeCost = def.upgrades[tower.level]
      ? def.upgrades[tower.level].upgradeCost
      : 0;
  }

  sellTower(tower) {
    const refund = Math.floor(tower.goldSpent * 0.8);
    this.game.gold += refund;
    this.towers = this.towers.filter(t => t !== tower);
    if (tower.spot) tower.spot.occupied = false;
  }

  drawTowers(ctx) {
    this.towers.forEach(t => {
      const drawRadius = 24 + t.level * 4;
      ctx.beginPath();
      ctx.arc(t.x, t.y, drawRadius, 0, Math.PI * 2);
      ctx.fillStyle = (t.type === "point") ? "blue" : "red";
      ctx.fill();
      ctx.strokeStyle = "#fff";
      ctx.stroke();

      // optional range circle
      ctx.beginPath();
      ctx.arc(t.x, t.y, t.range, 0, Math.PI * 2);
      ctx.strokeStyle = "rgba(255,255,255,0.3)";
      ctx.stroke();
    });
  }

  drawProjectiles(ctx) {
    ctx.fillStyle = "yellow";
    this.projectiles.forEach(proj => {
      ctx.fillRect(proj.x - 2, proj.y - 2, proj.w, proj.h);
    });
  }
}

File: python/tower_manager.py
import math
import pygame

class TowerManager:
    def __init__(self, game):
        self.game = game
        self.towers = []
        self.projectiles = []

        self.tower_types = [
            {
                "type": "point",
                "basePrice": 80,
                "range": 169,
                "splashRadius": 0,
                "fireRate": 1.5,
                "upgrades": [
                    { "level": 1, "damage": 10, "upgradeCost": 0   },
                    { "level": 2, "damage": 15, "upgradeCost": 50  },
                    { "level": 3, "damage": 20, "upgradeCost": 100 },
                    { "level": 4, "damage": 25, "upgradeCost": 150 },
                ],
            },
            {
                "type": "splash",
                "basePrice": 80,
                "range": 104,
                "splashRadius": 50,
                "fireRate": 1.5,
                "upgrades": [
                    { "level": 1, "damage": 8,  "upgradeCost": 0   },
                    { "level": 2, "damage": 12, "upgradeCost": 50  },
                    { "level": 3, "damage": 16, "upgradeCost": 100 },
                    { "level": 4, "damage": 20, "upgradeCost": 150 },
                ],
            },
        ]

    def get_tower_data(self):
        return self.tower_types

    def create_tower(self, tower_type_name, x, y, spot):
        definition = next((t for t in self.tower_types if t["type"] == tower_type_name), None)
        if not definition:
            return None

        lvl_data = definition["upgrades"][0]
        tower = {
            "type": definition["type"],
            "level": lvl_data["level"],
            "damage": lvl_data["damage"],
            "range": definition["range"],
            "splashRadius": definition["splashRadius"],
            "fireRate": definition["fireRate"],
            "fireCooldown": 0.0,
            "upgradeCost": definition["upgrades"][1]["upgradeCost"] if len(definition["upgrades"]) > 1 else 0,
            "maxLevel": len(definition["upgrades"]),
            "x": x,
            "y": y,
            "spot": spot,
        }
        self.towers.append(tower)
        return tower

    def update(self, delta_sec):
        for tower in self.towers:
            tower["fireCooldown"] -= delta_sec
            if tower["fireCooldown"] <= 0:
                self.fire_tower(tower)
                tower["fireCooldown"] = tower["fireRate"]

        to_remove = []
        for proj in self.projectiles:
            self.update_projectile(proj, delta_sec)
            if proj["hit"]:
                if proj["splashRadius"] > 0:
                    for enemy in self.game.enemies:
                        dx = enemy["x"] - proj["targetX"]
                        dy = enemy["y"] - proj["targetY"]
                        dist2 = dx*dx + dy*dy
                        if dist2 <= (proj["splashRadius"] ** 2):
                            if enemy == proj["mainTarget"]:
                                enemy["hp"] -= proj["damage"]
                            else:
                                enemy["hp"] -= (proj["damage"] / 2.0)
                else:
                    if proj["mainTarget"] in self.game.enemies:
                        proj["mainTarget"]["hp"] -= proj["damage"]
                to_remove.append(proj)

        self.projectiles = [p for p in self.projectiles if p not in to_remove]

    def update_projectile(self, proj, delta_sec):
        step = proj["speed"] * delta_sec
        dx = proj["targetX"] - proj["x"]
        dy = proj["targetY"] - proj["y"]
        dist = math.hypot(dx, dy)
        if dist <= step:
            proj["x"] = proj["targetX"]
            proj["y"] = proj["targetY"]
            proj["hit"] = True
        else:
            proj["x"] += (dx / dist) * step
            proj["y"] += (dy / dist) * step

    def fire_tower(self, tower):
        in_range_enemies = []
        for enemy in self.game.enemies:
            dx = enemy["x"] - tower["x"]
            dy = enemy["y"] - tower["y"]
            dist2 = dx*dx + dy*dy
            if dist2 <= (tower["range"]*tower["range"]):
                in_range_enemies.append(enemy)

        if not in_range_enemies:
            return

        target = in_range_enemies[0]
        proj = {
            "x": tower["x"],
            "y": tower["y"],
            "speed": 300,
            "damage": tower["damage"],
            "splashRadius": tower["splashRadius"],
            "mainTarget": target,
            "targetX": target["x"],
            "targetY": target["y"],
            "hit": False,
            "w": 4,
            "h": 4
        }
        self.projectiles.append(proj)

    def upgrade_tower(self, tower):
        definition = next((t for t in self.tower_types if t["type"] == tower["type"]), None)
        if not definition:
            return
        if tower["level"] >= len(definition["upgrades"]):
            return

        next_lvl_index = tower["level"]
        next_lvl_data = definition["upgrades"][next_lvl_index]
        cost = next_lvl_data["upgradeCost"]
        if self.game.gold < cost:
            return

        self.game.gold -= cost
        tower["level"] += 1
        tower["damage"] = next_lvl_data["damage"]
        if tower["level"] < tower["maxLevel"]:
            tower["upgradeCost"] = definition["upgrades"][tower["level"]]["upgradeCost"]
        else:
            tower["upgradeCost"] = 0

    def draw_towers(self, screen):
        for tower in self.towers:
            rad = 12 + tower["level"] * 2
            color = (0,0,255) if tower["type"] == "point" else (255,0,0)
            pygame.draw.circle(screen, color, (tower["x"], tower["y"]), rad, 0)
            pygame.draw.circle(screen, (255,255,255), (tower["x"], tower["y"]), rad, 1)

            if self.game.debug_mode:
                pygame.draw.circle(
                    screen, (255,255,255),
                    (tower["x"], tower["y"]),
                    tower["range"], 1
                )

    def draw_projectiles(self, screen):
        for proj in self.projectiles:
            rect = pygame.Rect(proj["x"] - 2, proj["y"] - 2, proj["w"], proj["h"])
            pygame.draw.rect(screen, (255, 255, 0), rect)
